inliniac/suricata
# DNP3 Support - v22.2

Previous PR: #2350 

Changes from last PR:
- Fix PPC compile issue (untested)
- Fix CentOS 5 compile issue - required refactoring of the floating point extraction code (https://github.com/jasonish/suricata/blob/dev-dnp3-v22.2-rebased/src/app-layer-dnp3-objects.c#L248)
- Removed notice logging.
- Fix bool return in DNP3HasEvents()
- Remove extra AppLayerDecoderEventsFreeEvents(&tx->decoder_events)
- Fix style issue with opening braces.
- Remove stale TODO comment.
- Use explicit types in detect-dnp3 parsing code; use existing string to integer conversion functions.
- Merge detect-engine-dnp3 into detect-dnp3.
- Move struct DetectDNP into detect-dnp3.c.
- Use a common JsonDNP3PrefixIsSize function.
- Remove output registration logging when no json support.
- Fix unit tests on OpenBSD (double mutex unlock)
- Cleanup func, iin, and obj registration as the list only needs to be registered once.

The diff prior to the rebase can be seen here:
https://gist.github.com/jasonish/6640aee31c7506330499803eb06fb69b

Mpm comment (https://github.com/inliniac/suricata/pull/2350#discussion_r84067158) has not been addressed.

This PR contains:
* An application decoder for DNP3.
* Snort compatible DNP3 keywords:
  * dnp3_data
  * dnp3_func
  * dnp3_ind
  * dnp3_obj
* A DNP3 output module for EVE (JSON logging).
* DNP3 support for Lua detection.

Documentation for the DNP3 rule keywords can be found here:
http://suricata.readthedocs.io/en/latest/rules/dnp3-keywords.html

Prscript:
- PR jasonish-pcap: https://buildbot.openinfosecfoundation.org/builders/jasonish-pcap/builds/25
- PR jasonish: https://buildbot.openinfosecfoundation.org/builders/jasonish/builds/375
